File: legacy/update-products.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
一键更新产品列表: 扫描 images/ 文件夹, 自动重新生成 products-data.js

用法: 双击 update-products.command, 或在终端运行 python3 update-products.py

文件命名规则 (放进 images/ 即可):
  home-7.png                      → HOME SERIES, 名字为占位符, tag 用默认值
  home-9_NAP TIME_cushion.png     → 自动填好名字 NAP TIME 和 tag cushion
  前缀对应分类:
    product- / art-   → ART
    home-             → HOME SERIES
    accessary- / accessory- → ACCESSORY
    handcraft-        → HANDCRAFT
    apparel-          → APPAREL
    petgoods-         → PET GOODS

已在 products-data.js 里手动改过的名字/tag 会自动保留, 放心运行。
"""
import os, re, json

ROOT = os.path.dirname(os.path.abspath(__file__))
IMG_DIR = os.path.join(ROOT, "images")
DATA = os.path.join(ROOT, "products-data.js")

# 前缀 → (数组名, 默认tag)
GROUPS = {
    "product":   ("ART_PRODUCTS", "print"),
    "art":       ("ART_PRODUCTS", "print"),
    "home":      ("HOME_PRODUCTS", "cushion"),
    "accessary": ("ACCESSORY_PRODUCTS", "charm"),
    "accessory": ("ACCESSORY_PRODUCTS", "charm"),
    "handcraft": ("HANDCRAFT_PRODUCTS", "handcraft"),
    "apparel":   ("APPAREL_PRODUCTS", "apparel"),
    "petgoods":  ("PETGOODS_PRODUCTS", "pet goods"),
}
ARRAYS = ["ART_PRODUCTS","HOME_PRODUCTS","ACCESSORY_PRODUCTS",
          "HANDCRAFT_PRODUCTS","APPAREL_PRODUCTS","PETGOODS_PRODUCTS"]
EXTS = (".png",".jpg",".jpeg",".webp")

# ---- 读取现有 products-data.js, 保留手动填过的 name/tag ----
existing = {}
if os.path.exists(DATA):
    txt = open(DATA, encoding="utf-8").read()
    for m in re.finditer(r'\{\s*img:"([^"]+)"\s*,\s*name:"([^"]*)"\s*,\s*tag:"([^"]*)"\s*\}', txt):
        existing[m.group(1)] = (m.group(2), m.group(3))

# ---- 扫描 images/ ----
# 同一编号存在多种格式时按此优先级取一张
EXT_PRIORITY = {".png":0, ".webp":1, ".jpg":2, ".jpeg":3}
found = {}   # (数组名, 编号) → (优先级, 文件名, name部分, tag部分)
skipped = []
for f in sorted(os.listdir(IMG_DIR)):
    ext = os.path.splitext(f)[1].lower()
    if ext not in EXT_PRIORITY: continue
    m = re.match(r'([a-zA-Z]+)-(\d+)(?:_([^_]+)_([^_.]+))?\.\w+$', f)
    if not m:
        skipped.append(f); continue
    prefix, num = m.group(1).lower(), int(m.group(2))
    if prefix not in GROUPS:
        skipped.append(f); continue
    arr, _ = GROUPS[prefix]
    key = (arr, num)
    cand = (EXT_PRIORITY[ext], f, m.group(3), m.group(4))
    if key not in found or cand[0] < found[key][0]:
        found[key] = cand

groups = {a: [] for a in ARRAYS}
for (arr, num), (_, f, fname_name, fname_tag) in found.items():
    prefix = f.split("-")[0].lower()
    default_tag = GROUPS[prefix][1]
    img = f"images/{f}"
    if img in existing:
        name, tag = existing[img]
    else:
        name = fname_name or "PRODUCT NAME"
        tag  = fname_tag or default_tag
    groups[arr].append((num, img, name, tag))

# ---- 写 products-data.js ----
out = ['/* ============================================================',
       '   产品数据 —— 由 update-products.py 自动生成',
       '   ------------------------------------------------------------',
       '   可以直接编辑 name / tag, 再次运行脚本时会保留你的修改。',
       '   新增图片: 放进 images/ 后运行 update-products.command 即可。',
       '   ============================================================ */','']
for arr in ARRAYS:
    out.append(f"const {arr} = [")
    for num, img, name, tag in sorted(groups[arr]):
        out.append(f'  {{ img:{json.dumps(img)}, name:{json.dumps(name)}, tag:{json.dumps(tag)} }},')
    out.append("];\n")
open(DATA, "w", encoding="utf-8").write("\n".join(out))

for arr in ARRAYS:
    print(f"{arr}: {len(groups[arr])} 件")
if skipped:
    print("未识别(需要 前缀-数字.png 命名):", ", ".join(skipped[:10]))
print("完成! products-data.js 已更新, 刷新网页即可。")

File: scripts/update-products.mjs
#!/usr/bin/env node

/**
 * Synchronize public/assets/products/* with data/products-data.js.
 *
 * Normal use:
 *   npm run products:update
 *
 * Preview without writing:
 *   npm run products:check
 *
 * Remove data entries whose image files no longer exist:
 *   npm run products:update -- --prune
 *
 * Existing id/name/tag/url values are preserved. New files may be named:
 *   accessory-magnet.png
 *   home-09__NAP TIME__cushion.png
 */

import fs from "node:fs";
import path from "node:path";
import { pathToFileURL } from "node:url";

const ROOT = path.resolve(import.meta.dirname, "..");
const PRODUCTS_ROOT = path.join(ROOT, "public", "assets", "products");
const DATA_FILE = path.join(ROOT, "data", "products-data.js");
const SHOULD_WRITE = !process.argv.includes("--check");
const SHOULD_PRUNE = process.argv.includes("--prune");

const CATEGORY_ORDER = [
  "accessory",
  "art",
  "home",
  "handcraft",
  "apparel",
  "doggoods"
];

const DEFAULT_TAGS = {
  accessory: "accessory",
  art: "print",
  home: "home goods",
  handcraft: "handcraft",
  apparel: "apparel",
  doggoods: "pet goods"
};

const IMAGE_EXTENSIONS = new Set([".png", ".jpg", ".jpeg", ".webp", ".avif", ".svg"]);
const IGNORED_NAMES = new Set([".DS_Store", "Thumbs.db", "coming_soon.png"]);

const source = fs.readFileSync(DATA_FILE, "utf8");
const moduleUrl = `${pathToFileURL(DATA_FILE).href}?updated=${Date.now()}`;
const { PRODUCTS: existingProducts } = await import(moduleUrl);

function normalizeWebPath(value) {
  return value.replaceAll("\\", "/");
}

function titleFromFilename(filename, category) {
  const stem = path.parse(filename).name;
  const metadata = stem.split("__");

  if (metadata.length >= 2 && metadata[1].trim()) {
    return metadata[1].trim();
  }

  const cleaned = stem
    .replace(new RegExp(`^(?:${category}|product)[-_]?`, "i"), "")
    .replace(/^\d+[-_]?/, "")
    .replace(/[-_]+/g, " ")
    .trim();

  return cleaned && !/^\d+$/.test(cleaned) ? cleaned.toUpperCase() : "PRODUCT NAME";
}

function tagFromFilename(filename, category) {
  const metadata = path.parse(filename).name.split("__");
  return metadata.length >= 3 && metadata[2].trim()
    ? metadata[2].trim()
    : DEFAULT_TAGS[category];
}

function slugify(value) {
  return value
    .toLowerCase()
    .normalize("NFKD")
    .replace(/[^\w\s-]/g, "")
    .trim()
    .replace(/[\s_]+/g, "-")
    .replace(/-+/g, "-");
}

function makeId(category, filename, usedIds) {
  const stem = path.parse(filename).name.split("__")[0];
  const number = stem.match(/(\d+)/)?.[1];
  const base = number
    ? `${category}-${number.padStart(2, "0")}`
    : `${category}-${slugify(stem.replace(new RegExp(`^${category}[-_]?`, "i"), "")) || "item"}`;

  let id = base;
  let suffix = 2;
  while (usedIds.has(id)) id = `${base}-${suffix++}`;
  usedIds.add(id);
  return id;
}

function scanCategory(category, currentItems) {
  const directory = path.join(PRODUCTS_ROOT, category);
  fs.mkdirSync(directory, { recursive: true });

  const currentByImage = new Map(currentItems.map(item => [item.image, item]));
  const usedIds = new Set(currentItems.map(item => item.id));
  const foundImages = [];

  for (const filename of fs.readdirSync(directory).sort((a, b) => a.localeCompare(b, "en", { numeric: true }))) {
    if (IGNORED_NAMES.has(filename) || filename.startsWith(".")) continue;
    if (!IMAGE_EXTENSIONS.has(path.extname(filename).toLowerCase())) continue;

    const image = normalizeWebPath(`/assets/products/${category}/${filename}`);
    foundImages.push(image);
  }

  const foundSet = new Set(foundImages);
  const missing = currentItems.filter(item => !foundSet.has(item.image));
  const keptMissing = SHOULD_PRUNE ? [] : missing;

  const synced = foundImages.map(image => {
    const current = currentByImage.get(image);
    if (current) return current;

    const filename = path.basename(image);
    return {
      id: makeId(category, filename, usedIds),
      image,
      name: titleFromFilename(filename, category),
      tag: tagFromFilename(filename, category),
      url: ""
    };
  });

  return {
    items: [...synced, ...keptMissing],
    added: synced.filter(item => !currentByImage.has(item.image)),
    missing
  };
}

function serializeProducts(products) {
  const lines = ["export const PRODUCTS = {"];

  CATEGORY_ORDER.forEach((category, categoryIndex) => {
    const items = products[category];
    lines.push(`  ${category}: [`);
    items.forEach((item, index) => {
      const comma = index === items.length - 1 ? "" : ",";
      lines.push(
        `    { id:${JSON.stringify(item.id)}, image:${JSON.stringify(item.image)}, ` +
        `name:${JSON.stringify(item.name)}, tag:${JSON.stringify(item.tag)}, url:${JSON.stringify(item.url || "")} }${comma}`
      );
    });
    lines.push(`  ]${categoryIndex === CATEGORY_ORDER.length - 1 ? "" : ","}`);
  });

  lines.push("};", "");
  return lines.join("\n");
}

const nextProducts = {};
let totalAdded = 0;
let totalMissing = 0;

for (const category of CATEGORY_ORDER) {
  const result = scanCategory(category, existingProducts[category] || []);
  nextProducts[category] = result.items;
  totalAdded += result.added.length;
  totalMissing += result.missing.length;

  console.log(`${category.padEnd(10)} ${String(result.items.length).padStart(2)} products`);
  result.added.forEach(item => console.log(`  + ${item.image}`));
  result.missing.forEach(item => {
    const action = SHOULD_PRUNE ? "removed from data" : "kept in data; use --prune to remove";
    console.warn(`  ! missing file: ${item.image} (${action})`);
  });
}

const categoriesStart = source.indexOf("export const CATEGORIES");
if (categoriesStart < 0) {
  throw new Error("Could not find `export const CATEGORIES` in data/products-data.js");
}

const nextSource = `${serializeProducts(nextProducts)}${source.slice(categoriesStart)}`;
const changed = nextSource !== source;

if (SHOULD_WRITE && changed) {
  fs.writeFileSync(DATA_FILE, nextSource, "utf8");
  console.log(`\nUpdated ${path.relative(ROOT, DATA_FILE)}.`);
} else if (!SHOULD_WRITE && changed) {
  console.log("\nChanges found. Run `npm run products:update` to write them.");
} else {
  console.log("\nProduct data is already up to date.");
}

console.log(`Added: ${totalAdded}; missing files: ${totalMissing}.`);

if (!SHOULD_WRITE && changed) process.exitCode = 1;
